pkg/source/openstack/client.go

```diff
@@ -236,13 +237,75 @@ func (c *Client) ExportVirtualMachine(vm *migration.VirtualMachineImport) error
 		return err
 	}
 
-	for vIndex, v := range vmObj.AttachedVolumes {
-		imageName := fmt.Sprintf("import-controller-%s-%d", vm.Spec.VirtualMachineName, vIndex)
+	// Helper function to do the export.
+	// This is necessary so that the defer functions are executed at the right
+	// time.
+	exportFn := func(index int, av servers.AttachedVolume) error {
+		var snapshot *snapshots.Snapshot
+		var volume *volumes.Volume
+		var volumeImage volumes.VolumeImage
+
+		imageName := fmt.Sprintf("import-controller-%s-%d", vm.Spec.VirtualMachineName, index)
+
+		// Make sure the snapshot, volume and volume image are cleaned up in any case.
+		defer func() {
```

Comment:
guess the `defer` needs to clean all resources which have been allocated from below loop:

```
for index, av := range vmObj.AttachedVolumes {
```

when any failure occurs

three slices:  snapshotList, volumeList, volumeImageList  are used to record them, and loop deleting

```
for _, volume := range volumeImageList {
// delete each
}
for ...
for ...

```

Reply:
Your suggestion is another solution on how this can be implemented. But honestly, I think the solution based on `defer` looks nicer and cleaner.

`defer` calls are executed when the function containing them exits. So we do not run into a situation where we have dozen of defers. The cleanup function contains 3 of them in summary, so we should not get in troubles here.